### PR TITLE
:bug: Add validation against cross-namespace secret references

### LINKED ISSUE #1965: Remove namespace field from secrets used by BMH API
**What steps did you take and what happened:**
The BareMetalHost CRD allows the userData, metaData, and networkData for
the provisioned host to be specified as links to k8s Secrets. There are
fields for both the Name and Namespace of the Secret, meaning that the
baremetal-operator will read a Secret from any namespace. A user with
access to create or edit a BareMetalHost can thus exfiltrate a Secret
from another namespace by using it as e.g. the userData for provisioning
some host (note that this need not be a real host, it could be a VM
somewhere).

**What did you expect to happen:**
We should fail with an error when a namespace is specified in the Secret
reference and it does not match the namespace of the BMH itself. We
should reject attempts to set such a configuration in the validating
webhook. And we should remove the namespace field in the next revision
of the API.

**Anything else you would like to add:**
https://github.com/metal3-io/baremetal-operator/security/advisories/GHSA-pqfh-xh7w-7h3p is limiting the secrets to same namespace now so API cleanup and validating webhook are just improvements/cleanup to the issue.

**Environment:**

- Baremetal Operator version: 0.8.0
- Environment (metal3-dev-env or other): any

@zaneb @dtantsur

### PR DESCRIPTION
**Pull request by mquhuy**
Add to BMH creation a validation, which checks if any secret references is from namespaces other than BMH's. This is to prevent the secret leakage when a user references secrets from other ns just to get the secrets out.

Fixes #1965

### CODE REVIEW

**Review by dtantsur — DISMISSED** [review]

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 1]
```diff
@@ -309,3 +314,34 @@ func validateRebootAnnotation(rebootAnnotation string) error {
...
+	if ref != nil &&
+		ref.Namespace != "" &&
+		ref.Namespace != hostNamespace {
+		return k8serrors.NewForbidden(
+			schema.GroupResource{
+				Group:    "metal3.io",
+				Resource: "baremetalhosts",
+			},
+			hostName,
+			fmt.Errorf("%s: cross-namespace Secret references are not allowed", fieldName),
```
Comment (tuminoid) [inline apis/metal3.io/v1alpha1/baremetalhost_validation.go:328]: Should we print the offending/expected namespace here?

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 2]
```diff
@@ -309,3 +314,34 @@ func validateRebootAnnotation(rebootAnnotation string) error {
 
 	return nil
 }
+
+func validateCrossNamespaceSecretReferences(hostNamespace, hostName, fieldName string, ref *corev1.SecretReference) error {
```
Comment (tuminoid) [inline apis/metal3.io/v1alpha1/baremetalhost_validation.go:318]: Document the function

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 3]
```diff
@@ -309,3 +314,34 @@ func validateRebootAnnotation(rebootAnnotation string) error {
...
+				Resource: "baremetalhosts",
+			},
+			hostName,
+			fmt.Errorf("%s: cross-namespace Secret references are not allowed", fieldName),
+		)
+	}
+	return nil
+}
+
+func (host *BareMetalHost) validateCrossNamespaceSecretReferences() []error {
```
Comment (tuminoid) [inline apis/metal3.io/v1alpha1/baremetalhost_validation.go:334]: Document the function.

File: apis/metal3.io/v1alpha1/baremetalhost_validation_test.go [hunk 4]
```diff
@@ -678,6 +679,97 @@ func TestValidateCreate(t *testing.T) {
...
+					MetaData: &corev1.SecretReference{
+						Name:      "test-secret3",
+						Namespace: "different-namespace3",
+					},
+				},
+			},
+			oldBMH:    nil,
+			wantedErr: "baremetalhosts.metal3.io \"test\" is forbidden: userData: cross-namespace Secret references are not allowed", // Should catch at least one error
+		},
+		{
```
Comment (tuminoid) [inline apis/metal3.io/v1alpha1/baremetalhost_validation_test.go:750]: Add a test where namespace is nil, if that is ever allowed/possible to be nil.

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go [hunk 5]
```diff
@@ -309,3 +314,34 @@ func validateRebootAnnotation(rebootAnnotation string) error {
 
 	return nil
 }
+
+func validateCrossNamespaceSecretReferences(hostNamespace, hostName, fieldName string, ref *corev1.SecretReference) error {
+	if ref != nil &&
+		ref.Namespace != "" &&
```
[the remark this answers is not in the document]
Comment (mquhuy) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:320]: Namespace is of type string, it's never nil.

File: apis/metal3.io/v1alpha1/baremetalhost_validation_test.go (hunk 4 above)
Comment (mquhuy) [reply apis/metal3.io/v1alpha1/baremetalhost_validation_test.go:750]: As said namespace cannot be nil.

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go (hunk 1 above)
Comment (mquhuy) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:328]: I don't think we should. And what would be the expected ns in this case? The secret's or the BMH's?

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go (hunk 2 above)
Comment (mquhuy) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:318]: Hmm. OK I will, but imo these functions are quite self-explanatory, and the other validate...() func in this files don't seem to have docstring either.

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go (hunk 1 above)
Comment (tuminoid) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:328]: Since there is only one allowed namespace, maybe it does not make too much sense. Printing the offending namespace wouldn't hurt IMO, but I'm fine either way.

File: apis/metal3.io/v1alpha1/baremetalhost_validation.go (hunk 2 above)
Comment (tuminoid) [reply apis/metal3.io/v1alpha1/baremetalhost_validation.go:320]: We are missing the linter for that, but want to enable it like in CAPM3, so no need to add future work. It is good to document them right away, it is the proper thing to do.

**Review by tuminoid — COMMENTED** [review]
/lgtm

**Review by kashifest — COMMENTED** [review]
/approve
